lib/scanner/xml-scanner/xml-scanner.c

```diff
@@ -0,0 +1,300 @@
+      {
+        return TRUE;
+      }
+
+  return FALSE;
+}
+
+static GMarkupParser skip = {};
+
+void
+xml_scanner_start_element_method(GMarkupParseContext  *context,
+                                 const gchar          *element_name,
+                                 const gchar         **attribute_names,
+                                 const gchar         **attribute_values,
+                                 gpointer              user_data,
+                                 GError              **error)
+{
+  XMLScanner *scanner = (XMLScanner *)user_data;
+
+  gchar *reversed = NULL;
+  guint tag_length = strlen(element_name);
+
+  if (scanner->matchstring_shouldreverse)
+    {
+      reversed = g_utf8_strreverse(element_name, tag_length);
+    }
+
+  if (tag_matches_patterns(scanner->exclude_patterns, tag_length, element_name, reversed))
+    {
+      msg_debug("xml: subtree skipped",
+                evt_tag_str("tag", element_name));
+      scanner->pop_next_time = 1;
+      g_markup_parse_context_push(context, &skip, NULL);
+      g_free(reversed);
+      return;
+    }
+
+  g_string_append_c(scanner->key, '.');
+  g_string_append(scanner->key, element_name);
+  scanner_push_attributes(scanner, scanner->key, attribute_names, attribute_values);
+
+  g_free(reversed);
+}
+
+static gint
+before_last_dot(GString *str)
+{
+  const gchar *s = str->str;
+  gchar *pos = strrchr(s, '.');
+  return (pos-s);
+}
+
+void
+xml_scanner_end_element_method(GMarkupParseContext *context,
+                               const gchar         *element_name,
+                               gpointer             user_data,
+                               GError              **error)
+{
+  XMLScanner *scanner = (XMLScanner *)user_data;
+
+  if (scanner->pop_next_time)
+    {
+      g_markup_parse_context_pop(context);
+      scanner->pop_next_time = 0;
+      return;
+    }
+  g_string_truncate(scanner->key, before_last_dot(scanner->key));
+  g_string_truncate(scanner->text, 0);
+}
+
+static void
+xml_scanner_strip_current_text(XMLScanner *self)
+{
+  g_strstrip(self->text->str);
+  g_string_set_size(self->text, strlen(self->text->str));
+}
+
+void
+xml_scanner_text_method(GMarkupParseContext *context,
+                        const gchar         *text,
+                        gsize                text_len,
+                        gpointer             user_data,
+                        GError             **error)
+{
+  if (text_len == 0)
+    return;
+
+  XMLScanner *scanner = (XMLScanner *)user_data;
+
+  g_string_append_len(scanner->text, text, text_len);
+  if (scanner->options->strip_whitespaces)
+    {
+      xml_scanner_strip_current_text(scanner);
+    }
+
+  if(scanner->text->len)
+    xml_scanner_push_current_key_value(scanner, scanner->key->str, scanner->text->str, scanner->text->len);
+}
+
+void
+xml_scanner_parse(XMLScanner *self, const gchar *input, gsize input_len, GError **error)
```

Comment:
What do you think about moving the construction and destruction of xml_ctx and scanner_callbacks to this function.

We could remove them from the XMLScanner structure, the scanner would be reusable this way, and it would fix our new/init problem as well.

Reply:
Good idea. :)
And of course it's needed if we would like to override virtual functions, e.g. `text_function` in case inheritance without loosing the current simple interface: an init() and a parse() function.